utils/utils.go

```diff
@@ -0,0 +1,44 @@
+package utils
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+)
+
+func RunAndWrapOnError(existingErr error, msg string, fn func() error) error {
+	if runErr := fn(); runErr != nil {
+		runErr = fmt.Errorf("%s: %v", msg, runErr)
+		if existingErr == nil {
+			return runErr
+		}
+		return fmt.Errorf("operation failed: %v, previous error: %w", runErr, existingErr)
+	}
+	return existingErr
+}
+
+func Cause(ctx context.Context) error {
```

Comment:
I love the comment. Will add.

I personally prefer the `Cause` name since `utils.Cause` vs. `context.Cause` should be enough to signal that something is different, while still being closely related.

Reply:
I think that's OK, but, devil's advocacy: it's a bit of a tease to signal that something is different, while witholding potential clue about _what_ is different. It's in the comment, but best to surface as much digestible info as early as possible.

I didn't really love `External`, but no other adjectives sprang to mind. `Injected`? `Manual`? `Imperative`?

Reply:
Really appreciate the push-back here.

> it's a bit of a tease to signal that something is different, while witholding potential clue about what is different. 

100% agree. However, in this case I believe the reason is subtle enough that any adjective risks obscuring the meaning even more. I'll move forward with the comment for now, but if you feel strongly, I'll change the name (another candidate is `UnderlyingCause`).